### PR TITLE
Adds fill paramter to utils.draw_bounding_boxes

### LINKED ISSUE #3280: Add fill parameter to utils.draw_bounding_boxes.
## 🚀 Feature

Fill parameter allows creating a semi-transparent box. This is particularly useful for Mask RCNN Model.
This would complete utils for Object detection and Instance Segmentation (least with rectangular boxes)

## Motivation

In Instance segmentation models, we also care about masks, not just the bounding box. Fill parameter allows us to fill in a semi-transparent way. Also, this parameter is optional hence it does not affect performance.

## Pitch

Add a param fill as follows
```
fill: Optional[List[Union[str, Tuple[int, int, int]]]] = None,

```

Here is complete running code with a few edits

```
@torch.no_grad()
def draw_bounding_boxes(
    image: torch.Tensor,
    boxes: torch.Tensor,
    labels: Optional[List[str]] = None,
    colors: Optional[List[Union[str, Tuple[int, int, int]]]] = None,
    fill: Optional[List[Union[str, Tuple[int, int, int]]]] = None,
    width: int = 1,
    font: Optional[str] = None,
    font_size: int = 10
) -> torch.Tensor:

    """
    Draws bounding boxes on given image.
    The values of the input image should be uint8 between 0 and 255.
    Args:
        image (Tensor): Tensor of shape (C x H x W)
        bboxes (Tensor): Tensor of size (N, 4) containing bounding boxes in (xmin, ymin, xmax, ymax) format. Note that
            the boxes are absolute coordinates with respect to the image. In other words: `0 <= xmin < xmax < W` and
            `0 <= ymin < ymax < H`.
        labels (List[str]): List containing the labels of bounding boxes.
        colors (List[Union[str, Tuple[int, int, int]]]): List containing the colors of bounding boxes. The colors can
            be represented as `str` or `Tuple[int, int, int]`.
        fill: Optional[List[Union[str, Tuple[int, int, int]]]] = None,
        width (int): Width of bounding box.
        font (str): A filename containing a TrueType font. If the file is not found in this filename, the loader may
            also search in other directories, such as the `fonts/` directory on Windows or `/Library/Fonts/`,
            `/System/Library/Fonts/` and `~/Library/Fonts/` on macOS.
        font_size (int): The requested font size in points.
    """

    if not isinstance(image, torch.Tensor):
        raise TypeError(f"Tensor expected, got {type(image)}")
    elif image.dtype != torch.uint8:
        raise ValueError(f"Tensor uint8 expected, got {image.dtype}")
    elif image.dim() != 3:
        raise ValueError("Pass individual images, not batches")

    ndarr = image.permute(1, 2, 0).numpy()
    img_to_draw = Image.fromarray(ndarr)

    img_boxes = boxes.to(torch.int64).tolist()

    draw = ImageDraw.Draw(img_to_draw, "RGBA")

    txt_font = ImageFont.load_default() if font is None else ImageFont.truetype(font=font, size=font_size)

    for i, bbox in enumerate(img_boxes):
        color = None if colors is None else colors[i]
        draw.rectangle(bbox, width=width, outline=color, fill=fill)

        if labels is not None:
            draw.text((bbox[0], bbox[1]), labels[i], fill=color, font=txt_font)

    return torch.from_numpy(np.array(img_to_draw)).permute(2, 0, 1)
```

This makes mask RCNN output more clear, and people can play with fill parameter such as confidence based fill, fill with colors different per class, etc.

## Additional context

I can send PR for this :sweat_smile:  I'm attaching outputs of above code.

draw_boxes_util2

(Sorry PyTorch logo :pray: )

draw_boxes_util

### PR DESCRIPTION
**Pull request by oke-aditya**
Closes #3280 .

Currently, I have kept fill to `Tuple[int, int, int, int]` or `str`.
Let me know if changes are needed.

Also, let me know how we can test this.

### CODE REVIEW

File: torchvision/utils.py [hunk 1]
```diff
@@ -143,6 +143,7 @@ def draw_bounding_boxes(
     boxes: torch.Tensor,
     labels: Optional[List[str]] = None,
     colors: Optional[List[Union[str, Tuple[int, int, int]]]] = None,
+    fill: Optional[List[Union[str, Tuple[int, int, int, int]]]] = None,
```
Comment (datumbox) [inline torchvision/utils.py:146]: I'm on the fence on this one while trying to balance flexibility and easy-of-use in this API. 

I think that having this as a list of colors would require people to provide too many parameters and pick too many colors. If you recall the palette size was one of the considerations on the first PR around this util, so adding the requirement to define even more colors might lead us to problems faster.

I think it's probably better to turn this into a boolean and fill the box with the same color as defined rather than supporting different colors for border and fill. Thoughts?

File: torchvision/utils.py [hunk 2]
```diff
@@ -151,7 +152,7 @@ def draw_bounding_boxes(
     """
     Draws bounding boxes on given image.
     The values of the input image should be uint8 between 0 and 255.
-
+    Resulting Tensor should be saved as PNG image.
```
Comment (datumbox) [inline torchvision/utils.py:155]: Could you add a space in between doc sections? There is another PR that is trying to fix the styling on documentation. :)

**Review by datumbox — COMMENTED** [review]
Thanks for the PR! I left a couple of comments for your consideration.

Also could you also change the unit-test to support the new param once you finalize the implementation? You will need to produce a new expected file but I think the code already supports this if you delete the old file.

File: torchvision/utils.py (hunk 1 above)
Comment (oke-aditya) [reply torchvision/utils.py:146]: Yes correct, color palette is an issue which rises quickly. 

Lets keep to boolean

Comment (oke-aditya) [reply torchvision/utils.py:146]: For semi transparent fill, the fill is `Tuple[int, int, int, int]` and not `Tuple[int, int, int]`. So what defaulot value should we keep for alpha channel.

File: torchvision/utils.py [hunk 3]
```diff
@@ -179,12 +183,21 @@ def draw_bounding_boxes(
 
     img_boxes = boxes.to(torch.int64).tolist()
 
-    draw = ImageDraw.Draw(img_to_draw)
+    if fill:
+        draw = ImageDraw.Draw(img_to_draw, "RGBA")
+        fill = colors + [100]
```
Comment (oke-aditya) [inline torchvision/utils.py:188]: I'm keeping `A` channel value fixed to 100. Otherwise we would have to make user pass `Tuple[int, int, int, int]`

File: torchvision/utils.py [hunk 4]
```diff
@@ -151,6 +152,7 @@ def draw_bounding_boxes(
     """
     Draws bounding boxes on given image.
     The values of the input image should be uint8 between 0 and 255.
+    If filled Resulting Tensor should be saved as PNG image.
```
Comment (oke-aditya) [inline torchvision/utils.py:155]: We need to use PNG only if It is filled.

File: torchvision/utils.py [hunk 5]
```diff
@@ -179,12 +183,21 @@ def draw_bounding_boxes(
...
+    else:
+        draw = ImageDraw.Draw(img_to_draw)
+
     txt_font = ImageFont.load_default() if font is None else ImageFont.truetype(font=font, size=font_size)
 
     for i, bbox in enumerate(img_boxes):
         color = None if colors is None else colors[i]
-        draw.rectangle(bbox, width=width, outline=color)
+        if fill:
+            fill_color = color + [100]
```
Comment (oke-aditya) [inline torchvision/utils.py:197]: I think we would need to set alpha channel to some hardcoded value. Since that gives the semi transparent look.

**Review by datumbox — APPROVED** [review]
Would you be able to update the following test to use your new feature?

https://github.com/pytorch/vision/blob/a0e5bbeab4b39ed1078384d10f742f757f74df00/test/test_utils.py#L83-L97

Feel free to update the expected file as well.

File: torchvision/utils.py (hunk 5 above)
Comment (datumbox) [reply torchvision/utils.py:197]: I just noticed that this causes issues on the [type checks](https://app.circleci.com/pipelines/github/pytorch/vision/5971/workflows/02e568f7-273c-4816-ba89-5dbe0a001d01/jobs/385457). Also this is not going to work for all values of `color`. There are various corner cases:
1. The `color` could be None
2. It could be a string
3. Or a tuple.

It might be worth confirming that all work as expected by adding a test, what you think?

**Review by datumbox — COMMENTED** [review]
@oke-aditya I added a couple of more comments, please let me know what you think.

**Review by datumbox — CHANGES_REQUESTED** [review]
.

File: torchvision/utils.py (hunk 5 above)
Comment (oke-aditya) [reply torchvision/utils.py:197]: If fill is true

If color is `None` lets fill with default color. 

If color is `Tuple[int, int, int]` I think we just add alpha of 100 as above and do. 

If color is `str` I think we need to get RGB value of the str add default alpha of 100 and draw ?

Comment (datumbox) [reply torchvision/utils.py:197]: I think we need to handle the `fill_color = color + [100]` especially in each corner case:

```
>>> None + [100]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for +: 'NoneType' and 'list'

>>> (0, 0, 0) + [100]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: can only concatenate tuple (not "list") to tuple

>>> "red" + [100]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: can only concatenate str (not "list") to str
```

Comment (oke-aditya) [reply torchvision/utils.py:197]: Sure. Sorry, I'm bit low on bandwidth this week but will complete by next week.
I will add tests too to check these corner cases.

Comment (datumbox) [reply torchvision/utils.py:197]: Don't worry mate. :) Take your time! Ping me when ready to make sure I review it again. Thanks!

**Comment by oke-aditya** [comment]
Done ! I just modified the test, which should now run with filled option. Since it covers all cases it should be fine.

**Review by datumbox — APPROVED** [review]
LGTM, thanks a lot for the PR @oke-aditya!
